### PR TITLE
Refactor auto rebase and auto squash

### LINKED ISSUE #1311: Make `RebaseCustom` more serialisable
Most of the backends now rely on `RebaseCustom` and `RebaseSquash` passes. Since these two passes take a circuit-to-circuit function (e.g. `basis_tk2_replacement`) as an argument, they can't be completely serialised. This behaviour makes some of our default passes non-(de)serialisable. However, since these circuit-to-circuit functions are usually built-in functions (e.g. `TK2_using_CX`) in tket, we should be able to make them more serialisable. One possible way is to override `RebaseCustom` and `RebaseSquash` to take an enum that points to one of the built-in functions.

Note that some TK1 transformations are implemented in python in `_decompositions.py`, which might make things trickier...

### PR DESCRIPTION
**Pull request by yao-cqc**
# Description

Move auto_rebase_pass and auto_squash_pass to C++ so they can have their own serialisations.

# Related issues

Should resolve #1311

# Checklist

- [x] I have performed a self-review of my code.
- [ ] I have commented hard-to-understand parts of my code.
- [x] I have made corresponding changes to the public API documentation.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have updated the changelog with any user-facing changes.

### CODE REVIEW

File: tket/src/Predicates/PassGenerators.cpp [hunk 1]
```diff
@@ -126,6 +126,169 @@ PassPtr gen_squash_pass(
...
+  j["basis_allowed"] = allowed_gates;
+  j["allow_swaps"] = allow_swaps;
+  return std::make_shared<StandardPass>(precons, t, pc, j);
+}
+
+PassPtr gen_auto_squash_pass(const OpTypeSet& singleqs) {
+  auto find_squash = [singleqs]() {
+    if (singleqs.find(OpType::Rz) != singleqs.end() &&
+        singleqs.find(OpType::PhasedX) != singleqs.end()) {
+      return Transforms::squash_1qb_to_Rz_PhasedX(true);
```
Comment (cqc-alec) [inline tket/src/Predicates/PassGenerators.cpp:274]: It's strange (but consistent with existing behaviour) that we set `always_squash_symbols` to `true` here but to `false` in line 277. I suggest making `false` here as well.

**Review by cqc-alec — COMMENTED** [review]
Looks good! Just some small comments.

File: tket/include/tket/Circuit/CircPool.hpp [hunk 2]
```diff
@@ -404,6 +404,9 @@ Circuit TK2_using_ZZPhase_and_swap(
  */
 Circuit TK2_using_TK2_or_swap(
     const Expr &alpha, const Expr &beta, const Expr &gamma);
+
+Circuit TK2_using_TK2(const Expr &alpha, const Expr &beta, const Expr &gamma);
```
[the remark this answers is not in the document]
Comment (yao-cqc) [reply tket/include/tket/Circuit/CircPool.hpp:409]: Done

File: tket/test/src/test_Rebase.cpp [hunk 3]
```diff
@@ -459,6 +459,13 @@ SCENARIO("Check cases for tk1_to_rzsx") {
       Transforms::remove_redundancies().apply(result);
       REQUIRE(result.n_gates() == test.expected_gates);
       REQUIRE(test_unitary_comparison(correct, result));
+      // allow x gates
+      Circuit result2 =
+          CircPool::tk1_to_rzxsx(test.alpha, test.beta, test.gamma);
+      Transforms::remove_redundancies().apply(result2);
+      REQUIRE(
+          result2.count_gates(OpType::X) ==
+          test.expected_gates - result2.n_gates());
```
[the remark this answers is not in the document]
Comment (yao-cqc) [reply tket/test/src/test_Rebase.cpp:471]: I've added a comment explaining this

File: tket/src/Predicates/PassGenerators.cpp [hunk 4]
```diff
@@ -126,6 +126,169 @@ PassPtr gen_squash_pass(
   return std::make_shared<StandardPass>(precons, t, postcon, j);
 }
 
+static std::function<Circuit(const Expr&, const Expr&, const Expr&)>
+find_tk1_replacement(const OpTypeSet& gateset) {
+  if (gateset.find(OpType::TK1) != gateset.end()) {
```
[the remark this answers is not in the document]
Comment (yao-cqc) [reply tket/src/Predicates/PassGenerators.cpp]: Thanks for pointing this out!

File: pytket/tests/passes_serialisation_test.py [hunk 5]
```diff
@@ -293,6 +295,7 @@ def nonparam_predicate_dict(name: str) -> Dict[str, Any]:
     "GreedyPauliSimp": standard_pass_dict(
         {"name": "GreedyPauliSimp", "discount_rate": 0.4, "depth_weight": 0.5}
     ),
+    "AutoSquash": standard_pass_dict({"name": "AutoSquash", "basis_singleqs": ["TK1"]}),
```
[the remark this answers is not in the document]
Comment (yao-cqc) [reply pytket/tests/passes_serialisation_test.py]: Done, just needed to make the order of the serialised lists deterministic.

**Review by Adrian999999999 — APPROVED** [review]
Hello

**Review by cqc-alec — APPROVED** [review]
Thanks!

**Pull request merged**